### ISSUE TITLE
Writing to Azure Storage Fails

### ISSUE THREAD

**Issue by trhumphries** [dotnet/dotnet-monitor#1338]
### Description
I have dotnet monitor running in Kubernetes when I try to dump the monitored process it returns a 202 however in the logs it shows an exception
The dump command and exception are:
/dump?type=Full&egressProvider=azureStorage

```JSONC
{
    "Timestamp": "2022-01-18T20:37:30.2657405Z",
    "EventId": 1,
    "LogLevel": "Error",
    "Category": "Microsoft.Diagnostics.Monitoring.WebApi.EgressOperation",
    "Message": "Request failed.",
    "Exception": "Microsoft.Diagnostics.NETCore.Client.ServerErrorException: WriteDumpAsync failed - HRESULT: 0x80004005    at Microsoft.Diagnostics.NETCore.Client.DiagnosticsClient.ValidateResponseMessage(IpcMessage responseMessage, String operationName, ValidateResponseOptions options)    at Microsoft.Diagnostics.NETCore.Client.DiagnosticsClient.WriteDumpAsync(DumpType dumpType, String dumpPath, Boolean logDumpGeneration, CancellationToken token)    at Microsoft.Diagnostics.Monitoring.WebApi.DumpService.DumpAsync(IEndpointInfo endpointInfo, DumpType mode, CancellationToken token) in /_/src/Microsoft.Diagnostics.Monitoring.WebApi/DumpService.cs:line 69    at Microsoft.Diagnostics.Tools.Monitor.Egress.AzureBlob.AzureBlobEgressProvider.EgressAsync(AzureBlobEgressProviderOptions options, Func\u00602 action, EgressArtifactSettings artifactSettings, CancellationToken token) in /_/src/Tools/dotnet-monitor/Egress/AzureBlob/AzureBlobEgressProvider.cs:line 48    at Microsoft.Diagnostics.Tools.Monitor.Egress.EgressService.EgressAsync(String providerName, Func\u00602 action, String fileName, String contentType, IEndpointInfo source, CancellationToken token) in /_/src/Tools/dotnet-monitor/Egress/EgressService.cs:line 67    at Microsoft.Diagnostics.Monitoring.WebApi.EgressOperation.\u003C\u003Ec__DisplayClass5_0.\u003C\u003CExecuteAsync\u003Eb__0\u003Ed.MoveNext() in /_/src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperation.cs:line 47 --- End of stack trace from previous location ---    at Microsoft.Diagnostics.Monitoring.WebApi.ExecutionHelper.InvokeAsync[T](Func\u00602 action, ILogger logger, CancellationToken token) in /_/src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs:line 48",
    /*
        Microsoft.Diagnostics.NETCore.Client.ServerErrorException: WriteDumpAsync failed - HRESULT: 0x80004005    at 
        Microsoft.Diagnostics.NETCore.Client.DiagnosticsClient.ValidateResponseMessage(IpcMessage responseMessage, String operationName, ValidateResponseOptions options)    at 
        Microsoft.Diagnostics.NETCore.Client.DiagnosticsClient.WriteDumpAsync(DumpType dumpType, String dumpPath, Boolean logDumpGeneration, CancellationToken token)    at 
        Microsoft.Diagnostics.Monitoring.WebApi.DumpService.DumpAsync(IEndpointInfo endpointInfo, DumpType mode, CancellationToken token) 
            in /_/src/Microsoft.Diagnostics.Monitoring.WebApi/DumpService.cs:line 69    at 
        Microsoft.Diagnostics.Tools.Monitor.Egress.AzureBlob.AzureBlobEgressProvider.EgressAsync(AzureBlobEgressProviderOptions options, Func\u00602 action, EgressArtifactSettings artifactSettings, CancellationToken token) 
            in /_/src/Tools/dotnet-monitor/Egress/AzureBlob/AzureBlobEgressProvider.cs:line 48    at 
        Microsoft.Diagnostics.Tools.Monitor.Egress.EgressService.EgressAsync(String providerName, Func\u00602 action, String fileName, String contentType, IEndpointInfo source, CancellationToken token) 
            in /_/src/Tools/dotnet-monitor/Egress/EgressService.cs:line 67    at 
        Microsoft.Diagnostics.Monitoring.WebApi.EgressOperation.\u003C\u003Ec__DisplayClass5_0.\u003C\u003CExecuteAsync\u003Eb__0\u003Ed.MoveNext() 
            in /_/src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperation.cs:line 47 
        --- End of stack trace from previous location ---    at 
        Microsoft.Diagnostics.Monitoring.WebApi.ExecutionHelper.InvokeAsync[T](Func\u00602 action, ILogger logger, CancellationToken token) 
            in /_/src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs:line 48
        
    */
    "State": {
        "Message": "Request failed.",
        "{OriginalFormat}": "Request failed."
    },
    "Scopes": [
        {
            "Message": "ArtifactType:dump ArtifactSource_ProcessId:1 ArtifactSource_RuntimeInstanceCookie:2a68b12d4c44458f8896dc83be57d0a9",
            "ArtifactType": "dump",
            "ArtifactSource_ProcessId": "1",
            "ArtifactSource_RuntimeInstanceCookie": "2a68b12d4c44458f8896dc83be57d0a9"
        }
    ]
}
```

### Configuration

My egress is defined as a ConfigMap:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: dotnetmonitor-egress 
  namespace: trial
data:
    Egress__AzureBlobStorage__azureStorage__accountUri: "https://trial.blob.core.windows.net"
    Egress__AzureBlobStorage__azureStorage__containerName: "trial-web-qa"
    Egress__AzureBlobStorage__azureStorage__blobPrefix: "highcpu"
    Egress__AzureBlobStorage__azureStorage__accountKeyName: "MonitorBlobAccountKey"
    Egress__Properties__MonitorBlobAccountKey: "R2JuTUHV/bjyw=="
```

and my deployment is:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: trial-web-trh
  namespace: trial
  labels:
    environment: QA
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: trial-web-trh
  template:
    metadata:
      labels:
        app.kubernetes.io/name: trial-web-trh
    spec:
      containers:
        - name: trial-web
          image: mycrl.azurecr.io/trial-web:5.0.5.3
          command:
            - dotnet
            - Trail.Web.dll
          ports:
            - containerPort: 443
              protocol: TCP
          env:
            - name: ASPNETCORE_ENVIRONMENT
              value: CI
            - name: DOTNET_DiagnosticPorts
              value: /diag/port
          volumeMounts:
          - mountPath: /diag
            name: diagvol
          - mountPath: /dumps
            name: dumpsvol
          readinessProbe:
            httpGet:
              path: /health
              port: 443
              scheme: HTTPS
            initialDelaySeconds: 10
            timeoutSeconds: 1
            periodSeconds: 15
            successThreshold: 1
            failureThreshold: 3
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: IfNotPresent
        - name: monitor
          image: mcr.microsoft.com/dotnet/monitor
          ports:
            - containerPort: 52323
          # DO NOT use the --no-auth argument for deployments in production; this argument is used for demonstration
          # purposes only in this example. Please continue reading after this example for further details.
          args: [ "--no-auth" ]
          imagePullPolicy: Always
          env:
            - name: DOTNETMONITOR_DiagnosticPort__ConnectionMode
              value: Listen
            - name: DOTNETMONITOR_DiagnosticPort__EndpointName
              value: /diag/port
            - name: DOTNETMONITOR_Urls
              value: "http://*:52323"
          volumeMounts:
            - mountPath: /etc/dotnet-monitor
              name: config
            - mountPath: /diag
              name: diagvol
            - mountPath: /dumps
              name: dumpsvol
      volumes:
      - name: config 
        projected:
          sources:
          - configMap:
              name: dotnetmonitor-egress
          - configMap:
              name: dotnetmonitor-high-cpu
          - configMap:
              name: dotnetmonitor-process
      - name: diagvol
        emptyDir: {}
      - name: dumpsvol
        emptyDir: {}
      restartPolicy: Always
      terminationGracePeriodSeconds: 30
      nodeSelector:
        kubernetes.io/os: linux
```

### Other information

I can use the trace api and it does write a net trace file to Azure Storage however the file size is 0

> [2022-01-18] Edited by @kelltrick to use code blocks for easier reading.

**Comment by kelltrick** [comment]
Hi @trhumphries! First off, thanks the detailed bug report, I wouldn't have been able to make any headway without the provided Kubernetes deployment file. This is a bit of a multi-faced issue, with some things being conjecture based on a repo I was able to put together based on the above deployment.

>tl;dr:
>Making some assumptions that are explained in greater detail below, try adding the security capability `SYS_PTRACE` to _your_ container.
>```yaml
>    spec:
>      containers:
>        - name: trial-web
>          securityContext:
>            capabilities:
>              add:
>              - SYS_PTRACE
>```

The first issue is that the error code you see (`Microsoft.Diagnostics.NETCore.Client.ServerErrorException: WriteDumpAsync failed - HRESULT: 0x80004005`) is a generic error result we get from the [runtime](https://github.com/dotnet/runtime) whenever dump creation fails for any reason. I spoke with @tommcdon and we want to pass through better diagnostic info for when `createdump` fails. I created dotnet/runtime#64069 to track the improvement in the runtime. Here are the details for the curious: `0x80004005` is [`DS_IPC_E_FAIL`](https://github.com/dotnet/runtime/blob/d0e2dac3a828b6afda5cff812b7b09b9dba213b7/src/native/eventpipe/ds-types.h#L132) from [`static ds_ipc_result_t ds_rt_generate_core_dump (DiagnosticsDumpCommandId commandId, DiagnosticsGenerateCoreDumpCommandPayload *payload)`](https://github.com/dotnet/runtime/blob/27fdc2d22f8717c7b0ba254ae9ef2900d00f5e7a/src/coreclr/vm/eventing/eventpipe/ds-rt-coreclr.h#L191). `ds_rt_generate_core_dump` will return `DS_IPC_E_FAIL` whenever the `createdump` process fails for any reason.

I hypothesize that you're encountering the same error I'm getting when I setup a slightly adapted version of your Kubernetes 
 cluster. However, we can't confirm this without some additional diagnostics. The error I encountered is a new way to encounter dotnet/diagnostics#1683 which is an overly complex "Permission Denied" error. What is happening in my case is that Kubernetes is running my process as "root" (which is the expected default) but Kubernetes always does this with a "runasuser" redirection that causes the process to be running with an effective user ID of `0/root` but with a different actual user ID. One of few things on Linux that makes the distinction between the 2 different types of user IDs is when protecting the pipes under `/proc/XXX/mem` (where `XXX` is the PID of a process running as `root` ). So, you end up in a scenario where your app is running as `root` but the separate process that spawns (in this case `createdump`) is also running as `root` but because of the different actual user ID, it can't open `/proc/1/mem` (assuming your target dotnet process is PID=1).

The steps to try next are as follows:
You can get the actual error message that's causing the error by requesting the dump manually.
1. Use kubectl to get into one of your pods (specifically we want to be on the container running the target app): 
```bash
kubectl exec --stdin --tty deployment/trial-web-trh --container trial-web -- /bin/bash
```
2. Launch `createdump` manually (you'll need to know the PID of your dotnet app, but it is probably `1`)
```bash
/usr/share/dotnet/shared/Microsoft.NETCore.App/6.0.1/createdump 1 --full --name /dumps/trial-web.1.nettrace --diag --verbose --crashreport
```
>Here is the [documentation](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/debug-linux-dumps) for `createdump`. The more detailed parameters (like `--verbose` and `--crashreport`) are used by the runtime [here](https://github.com/dotnet/runtime/blob/27fdc2d22f8717c7b0ba254ae9ef2900d00f5e7a/src/coreclr/pal/src/thread/process.cpp#L3130).
3. If you are hitting the same error as I got in my cluster, then I would expect the output of the `createdump` to look like this:
```
open(/proc/1/mem) FAILED 13 (Permission denied)
```
4. If the output of `createdump` matches what is shown in step 3, you may want to try adding `SYS_PTRACE` to the deployment, see `securityContext` below.
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: trial-web-trh
  namespace: trial
  labels:
    environment: QA
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: trial-web-trh
  template:
    metadata:
      labels:
        app.kubernetes.io/name: trial-web-trh
    spec:
      containers:
        - name: trial-web
          securityContext:
           capabilities:
             add:
             - SYS_PTRACE
          image: mycrl.azurecr.io/trial-web:5.0.5.3
          command:
            - dotnet
            - Trail.Web.dll
          ports:
            - containerPort: 443
              protocol: TCP
          env:
            - name: ASPNETCORE_ENVIRONMENT
              value: CI
            - name: DOTNET_DiagnosticPorts
              value: /diag/port
          volumeMounts:
          - mountPath: /diag
            name: diagvol
          - mountPath: /dumps
            name: dumpsvol
          readinessProbe:
            httpGet:
              path: /health
              port: 443
              scheme: HTTPS
            initialDelaySeconds: 10
            timeoutSeconds: 1
            periodSeconds: 15
            successThreshold: 1
            failureThreshold: 3
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: IfNotPresent
        - name: monitor
          image: mcr.microsoft.com/dotnet/monitor
          ports:
            - containerPort: 52323
          # DO NOT use the --no-auth argument for deployments in production; this argument is used for demonstration
          # purposes only in this example. Please continue reading after this example for further details.
          args: [ "--no-auth" ]
          imagePullPolicy: Always
          env:
            - name: DOTNETMONITOR_DiagnosticPort__ConnectionMode
              value: Listen
            - name: DOTNETMONITOR_DiagnosticPort__EndpointName
              value: /diag/port
            - name: DOTNETMONITOR_Urls
              value: "http://*:52323"
          volumeMounts:
            - mountPath: /etc/dotnet-monitor
              name: config
            - mountPath: /diag
              name: diagvol
            - mountPath: /dumps
              name: dumpsvol
      volumes:
      - name: config 
        projected:
          sources:
          - configMap:
              name: dotnetmonitor-egress
          - configMap:
              name: dotnetmonitor-high-cpu
          - configMap:
              name: dotnetmonitor-process
      - name: diagvol
        emptyDir: {}
      - name: dumpsvol
        emptyDir: {}
      restartPolicy: Always
      terminationGracePeriodSeconds: 30
      nodeSelector:
        kubernetes.io/os: linux
```
5. Try `createdump` again (steps 1 and 2). After applying the fix to my pod, the output of my `createdump` looked like this:
```bash
[... several more lines omitted...]
00007f73118e2000 - 00007f73118e3000 (000001) 0000000000000000 rw---b 46
00007f73118e3000 - 00007f73118ea000 (000007) 0000000000000000 r----b 44
00007f73118ea000 - 00007f73118fa000 (000010) 0000000000000000 r-x--b 45
00007f73118fa000 - 00007f7311900000 (000006) 0000000000000000 r----b 44
00007f7311900000 - 00007f7311907000 (000007) 0000000000000000 rw---b 46
00007f7311907000 - 00007f731190a000 (000003) 0000000000000000 r----b 44
00007f731190a000 - 00007f731192a000 (000020) 0000000000000000 r-x--b 45
00007f731192a000 - 00007f7311932000 (000008) 0000000000000000 r----b 44
00007f7311933000 - 00007f7311934000 (000001) 0000000000000000 r----b 44
00007f7311934000 - 00007f7311936000 (000002) 0000000000000000 rw---b 46
00007ffcddb7d000 - 00007ffcddb9e000 (000021) 0000000000000000 rw---b 46
00007ffcddbf8000 - 00007ffcddbf9000 (000001) 0000000000000000 r-x--b 45
ffffffffff600000 - ffffffffff601000 (000001) 0000000000000000 r-x--b 45
Writing crash report to file /dumps/trial-web.1.nettrace.crashreport.json
Writing full dump to file /dumps/trial-web.1.nettrace
Writing memory region headers to core file
Writing process information to core file
Writing 20 auxv entries to core file
Writing 470 NT_FILE entries to core file
Writing 21 thread entries to core file
Writing 581 memory regions to core file
Written 265220096 bytes (64751 pages) to core file
Dump successfully written
```

Let me know if this helps!

**Comment by kelltrick** [comment]
Can you also confirm that you have `Storage.DumpTempFolder` set to `/dumps`? It'll look something like this
```yaml
Storage__DumpTempFolder: "/dumps"
```

**Comment by kelltrick** [comment]
@wiktork you've investigated this area a lot more than me. Do you know why we might still get a Permission denied even with `SYS_PTRACE` permission? I'm curious if this is the root error that dotnet-monitor is getting.

**Comment by kelltrick** [comment]
For the image `mycrl.azurecr.io/trial-web:5.0.5.3` can you provided the `Dockerfile` used to build it? I'd need to know what image `trial-web` is based on and how it's configured to further speculate.

**Comment by kelltrick** [comment]
@trhumphries , feel free to re-open this if you get back to diagnosing this issue.

I'm closing this due to a lack of information to further diagnose the issue.
